core/src/main/java/com/linecorp/armeria/client/Endpoint.java

```diff
@@ -518,7 +526,22 @@ public Endpoint withWeight(int weight) {
         if (this.weight == weight) {
             return this;
         }
-        return new Endpoint(host(), ipAddr(), port, weight, hostType);
+        return new Endpoint(host(), ipAddr(), port, weight, hostType, metadata);
+    }
+
+    /**
+     * Returns a new host endpoint with the specified metadata.
+     *
+     * @return the new endpoint with the specified metadata. {@code this} if this endpoint has the same
+     *      metadata.
+     *
+     * @throws IllegalStateException if this endpoint is not a host but a group
+     */
+    public Endpoint withMetadata(AttributeMap metadata) {
```

Comment:
How about providing a way to set and get attributes rather than taking `AttributeMap`?
```java
T attr(AttributeKey<T> key)
T setAttr(AttributeKey<T> key, @Nullable T value);
void setAttrs(Iterable<? extends Entry<AttributeKey<T>, T>> attributes)
```

Reply:
Yes, It is much better. The code is updated.

Reply:
Had a chat with other maintainers about this feature.
`Endpoint` is an immutable object. Even if we provide metadata, we want to keep the immutability. 
Sorry for the wrong direction. How about creating a new Endpoint whenever new attributes are added.
```java
class Endpoint {

    <T> public Endpoint withAttr(AttributeKey<T> key, @Nullable T value) {
	    // create new AttributeMap by copy old attributes and add new value and;
        // return a new Endpoint
    }

    public Endpoint withAttrs(Iterable<? extends Entry<AttributeKey<?>, ?>> attributes) {
	    // create new AttributeMap by copy old attributes and add new values and;
        // return a new Endpoint
    }

    <T> T attr(AttributeKey<T> key) {
    }
}
```
In addition, we prefer to use `DefaultAttributeMap`. Could you move `DefaultAttributeMap` to `internal` package and use it for metadata?

Reply:
Sorry for the late update. The code is updated now.